File: lib/leetcode/254_factor-combinations.js
var getFactors = function(n) {
  const ans = [];

  const helper = (sofar, n, start) => {
    if (n <= 1) {
      if (sofar.length > 1) {
        ans.push(sofar.concat()); // a clone of sofar array
        return;
      } else {
        return [];
      }
    }

    for (var i = start; i <= n; i++) {
      // n = 8, 2,
      if (n % i === 0) {
        sofar.push(i); // [2, 2]
        helper(sofar, n / i, i);
        sofar.pop();
      }
    }
  }; // helper

  helper([], n, 2);
  return ans;
};

File: lib/leetcode/254_factor-combinations.rb
# @param {Integer} n
# @return {Integer[][]}
def get_factors(n)
  result = []

  factor_helper(result, [], n, 2)

  result
end

def factor_helper(result, sofar, target, start)
  if target <= 1
    result.push sofar.dup if sofar.size > 1
    return
  end

  start.upto(target) do |num|
    next unless target % num == 0

    sofar.push num
    factor_helper(result, sofar, target / num, num)
    sofar.pop
  end
end
#
# [
#   [2, 6],
#   [2, 2, 3],
#   [3, 4]
# ]
p get_factors(12)
p get_factors(25)
p get_factors(13)
p get_factors(1)
